### PR TITLE
Add T5 to delta modifier map

### PR DESCRIPTION
**Pull request by aaronrmm**
This adds an entry in [utils.modeling.py::MOIDIFIED_MODULES_DICT](https://github.com/CarperAI/trlx/blob/84a07113d8ced1497c5a7941a02f9c534bb4e185/trlx/utils/modeling.py#L279) for T5 as per
https://github.com/CarperAI/trlx/issues/173#issuecomment-1400497548

This is a guess using other modules as examples and by visualizing a T5 model with bigmodelvis.Visualization:
```
from bigmodelvis import Visualization
from transformers import AutoModelForSeq2SeqLM
backbone_model = AutoModelForSeq2SeqLM.from_pretrained("google/flan-t5-xl")
model_vis = Visualization(backbone_model)
graph = model_vis.structure_graph()
print(graph)
```
I can run it, though have not been able to do a full training job without running out of memory.

`print(graph)`output:

root
├── shared(Embedding),lm_head(Linear) weight:[32128, 2048]
├── encoder (T5Stack)
│   ├── embed_tokens (Embedding) weight:[32128, 2048]
│   ├── block (ModuleList)
│   │   ├── 0 (T5Block)
│   │   │   └── layer (ModuleList)
│   │   │       ├── 0 (T5LayerSelfAttention)
│   │   │       │   ├── SelfAttention (T5Attention)
│   │   │       │   │   ├── q,k,v,o(Linear) weight:[2048, 2048]
│   │   │       │   │   └── relative_attention_bias (Embedding) weight:[32, 32]
│   │   │       │   └── layer_norm (T5LayerNorm) weight:[2048]
│   │   │       └── 1 (T5LayerFF)
│   │   │           ├── DenseReluDense (T5DenseGatedActDense)
│   │   │           │   ├── wi_0,wi_1(Linear) weight:[5120, 2048]
│   │   │           │   └── wo (Linear) weight:[2048, 5120]
│   │   │           └── layer_norm (T5LayerNorm) weight:[2048]
│   │   └── 1-23(T5Block)
│   │       └── layer (ModuleList)
│   │           ├── 0 (T5LayerSelfAttention)
│   │           │   ├── SelfAttention (T5Attention)
│   │           │   │   └── q,k,v,o(Linear) weight:[2048, 2048]
│   │           │   └── layer_norm (T5LayerNorm) weight:[2048]
│   │           └── 1 (T5LayerFF)
│   │               ├── DenseReluDense (T5DenseGatedActDense)
│   │               │   ├── wi_0,wi_1(Linear) weight:[5120, 2048]
│   │               │   └── wo (Linear) weight:[2048, 5120]
│   │               └── layer_norm (T5LayerNorm) weight:[2048]
│   └── final_layer_norm (T5LayerNorm) weight:[2048]
└── decoder (T5Stack)
    ├── embed_tokens (Embedding) weight:[32128, 2048]
    ├── block (ModuleList)
    │   ├── 0 (T5Block)
    │   │   └── layer (ModuleList)
    │   │       ├── 0 (T5LayerSelfAttention)
    │   │       │   ├── SelfAttention (T5Attention)
    │   │       │   │   ├── q,k,v,o(Linear) weight:[2048, 2048]
    │   │       │   │   └── relative_attention_bias (Embedding) weight:[32, 32]
    │   │       │   └── layer_norm (T5LayerNorm) weight:[2048]
    │   │       ├── 1 (T5LayerCrossAttention)
    │   │       │   ├── EncDecAttention (T5Attention)
    │   │       │   │   └── q,k,v,o(Linear) weight:[2048, 2048]
    │   │       │   └── layer_norm (T5LayerNorm) weight:[2048]
    │   │       └── 2 (T5LayerFF)
    │   │           ├── DenseReluDense (T5DenseGatedActDense)
    │   │           │   ├── wi_0,wi_1(Linear) weight:[5120, 2048]
    │   │           │   └── wo (Linear) weight:[2048, 5120]
    │   │           └── layer_norm (T5LayerNorm) weight:[2048]
    │   └── 1-23(T5Block)
    │       └── layer (ModuleList)
    │           ├── 0 (T5LayerSelfAttention)
    │           │   ├── SelfAttention (T5Attention)
    │           │   │   └── q,k,v,o(Linear) weight:[2048, 2048]
    │           │   └── layer_norm (T5LayerNorm) weight:[2048]
    │           ├── 1 (T5LayerCrossAttention)
    │           │   ├── EncDecAttention (T5Attention)
    │           │   │   └── q,k,v,o(Linear) weight:[2048, 2048]
    │           │   └── layer_norm (T5LayerNorm) weight:[2048]
    │           └── 2 (T5LayerFF)
    │               ├── DenseReluDense (T5DenseGatedActDense)
    │               │   ├── wi_0,wi_1(Linear) weight:[5120, 2048]
    │               │   └── wo (Linear) weight:[2048, 5120]
    │               └── layer_norm (T5LayerNorm) weight:[2048]
    └── final_layer_norm (T5LayerNorm) weight:[2048]

### CODE REVIEW

**Review by jon-tow — CHANGES_REQUESTED** [review]
Thanks for looking into this @aaronrmm !

Right now, no LoRA adapters are injected into the model:

image

We'll need a new regex pattern for capturing T5 layers as they're not in the same format as the rest of the causal models we currently have. See the existing regex pattern below:
https://github.com/CarperAI/trlx/blob/b2ce1a4c17b16dba30c3174f7b479891d622bab6/trlx/utils/modeling.py#L334-L347

Essentially, we want a pattern that "walks" the decoder blocks with access to the linear layers within. Note: decoder-only because it's the only learnable part of our seq2seq models; see https://github.com/CarperAI/trlx/blob/b2ce1a4c17b16dba30c3174f7b479891d622bab6/trlx/utils/modeling.py#L47-L66

File: tests/test_utils.py [hunk 1]
```diff
@@ -97,6 +97,8 @@ def test_hf_attr_getters(model_name: str):
         "EleutherAI/gpt-neox-20b",
         "facebook/opt-1.3b",
         "bigscience/bloom-560m",
+        "google/flan-t5-large",
+        "google/flan-t5-xxl"
```
[the remark this answers is not in the document]
Comment (aaronrmm) [reply tests/test_utils.py]: @jon-tow 
@ethankim00 helped me with the layer naming we changed it to:
```
        "all": [
            "layer.0.SelfAttention.q",
            "layer.0.SelfAttention.k",
            "layer.0.SelfAttention.v",
            "layer.0.SelfAttention.o",
            "layer.1.EncDecAttention.q",
            "layer.1.EncDecAttention.k",
            "layer.1.EncDecAttention.v",
            "layer.1.EncDecAttention.o",
            "layer.1.DenseReluDense.wo",
            "layer.1.DenseReluDense.wi_0",
            "layer.1.DenseReluDense.wi_1",
            "layer.2.DenseReluDense.wo",
            "layer.2.DenseReluDense.wi_0",
            "layer.2.DenseReluDense.wi_1",
        ],
```
This is working in that I'm now getting more than zero parameters for those OpenDelta ratios. 

But additionally, are you saying we should exclude encoder layers from this modified modules list? And that would require some changes to the regex pattern.

[the remark this answers is not in the document]
Comment (jon-tow) [reply tests/test_utils.py]: The behavior for the causal (decoder-only) models is that only the upper trainable layers, determined by `num_layers_unfrozen`, are tagged with LoRA layers. If we want to be consistent, then for seq2seq models we need to also inject LoRA layers only into the trainable parts - the upper blocks of the decoder (see `freeze_bottom_seq2seq_layers()` above). This map seems to also modify layers in the encoder block. I think you may be able to just add a condition to modify the `base_model.decoder` for seq2seq archs and just `base_model` for causal in the line below.
https://github.com/CarperAI/trlx/blob/b2ce1a4c17b16dba30c3174f7b479891d622bab6/trlx/trainer/accelerate_base_trainer.py#L146

[the remark this answers is not in the document]
Comment (aaronrmm) [reply tests/test_utils.py]: Hmm. I am caught up on context and pushed a potential solution, but your suggestion might be better. I just don't fully understand. Let me know if you want to discuss. @jon-tow

[the remark this answers is not in the document]
Comment (jon-tow) [reply tests/test_utils.py]: @aaronrmm This is perfect! I wasn't sure if we'd need to modify the call in the base trainer but you proved it unnecessary :)

**Review by jon-tow — APPROVED** [review]
Thanks, @aaronrmm!

**Pull request merged**